### PR TITLE
Feat/add 0x47 SELFBALANCE opcode

### PR DESCRIPTION
**Pull request by ftupas**
## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [x] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: #87
Fixes: #87

## What is the new behavior?

- Format and rearrange Block Information opcodes accordingly
- 0x47 SELFBALANCE opcode implemented
- Get balance of currently executing contract by getting it from the execution context and querying its balance from the eth contract
- Add distinction among test cases for `test_execute` and `test_execute_at_address`

### CODE REVIEW

**Comment by LucasLvy** [comment]
The current contract address will most likely be passed in a call context. The caller address won’t be the self address

**Review by LucasLvy — CHANGES_REQUESTED** [review]
Format and rebase please

**Comment by ftupas** [comment]
> The current contract address will most likely be passed in a call context. The caller address won’t be the self address

Thanks for your feedback @LucasLvy 🤝 Got it let me change that

File: src/kakarot/execution_context.cairo [hunk 1]
```diff
@@ -32,7 +32,9 @@ namespace ExecutionContext {
         pedersen_ptr: HashBuiltin*,
         range_check_ptr,
         bitwise_ptr: BitwiseBuiltin*,
-    }(code: felt*, code_len: felt, calldata: felt*) -> model.ExecutionContext* {
+    }(
+        current_contract_address: Uint256*, code: felt*, code_len: felt, calldata: felt*
```
Comment (LucasLvy) [inline src/kakarot/execution_context.cairo:36]: Ethereum address is 20 bytes and starknet address is a felt so they both fit in a felt
```suggestion
        current_contract_sn_address: felt, code: felt*, code_len: felt, calldata: felt*
```

File: src/kakarot/execution_context.cairo [hunk 2]
```diff
@@ -46,6 +48,7 @@ namespace ExecutionContext {
         let memory: model.Memory* = Memory.init();
```
Comment (LucasLvy) [inline src/kakarot/execution_context.cairo:49]:
```suggestion
map current starknet address to evm address
```

File: src/kakarot/execution_context.cairo [hunk 3]
```diff
@@ -46,6 +48,7 @@ namespace ExecutionContext {
...
 
         local ctx: model.ExecutionContext* = new model.ExecutionContext(
+            current_contract_address=current_contract_address,
```
Comment (LucasLvy) [inline src/kakarot/execution_context.cairo:51]:
```suggestion
            self_sn_address=current_contract_sn_address,
            self_evm_address=mapped_sn_address,
```

File: src/kakarot/execution_context.cairo [hunk 4]
```diff
@@ -71,6 +74,7 @@ namespace ExecutionContext {
         let intrinsic_gas_cost = Constants.TRANSACTION_INTRINSIC_GAS_COST;
         let gas_used = self.gas_used + intrinsic_gas_cost;
         return new model.ExecutionContext(
+            current_contract_address=self.current_contract_address,
```
Comment (LucasLvy) [inline src/kakarot/execution_context.cairo:77]:
```suggestion
            self_sn_address=self.self_sn_address,
            self_evm_address=self.self_evm_address,
```

File: src/kakarot/execution_context.cairo [hunk 5]
```diff
@@ -106,6 +110,7 @@ namespace ExecutionContext {
         bitwise_ptr: BitwiseBuiltin*,
     }(self: model.ExecutionContext*) -> model.ExecutionContext* {
         return new model.ExecutionContext(
+            current_contract_address=self.current_contract_address,
```
Comment (LucasLvy) [inline src/kakarot/execution_context.cairo:113]:
```suggestion
            self_sn_address=self.self_sn_address,
            self_evm_address=self.self_evm_address,
```

File: src/kakarot/execution_context.cairo [hunk 6]
```diff
@@ -153,6 +158,7 @@ namespace ExecutionContext {
         bitwise_ptr: BitwiseBuiltin*,
     }(self: model.ExecutionContext*, new_stack: model.Stack*) -> model.ExecutionContext* {
         return new model.ExecutionContext(
+            current_contract_address=self.current_contract_address,
```
Comment (LucasLvy) [inline src/kakarot/execution_context.cairo:161]:
```suggestion
            self_sn_address=self.self_sn_address,
            self_evm_address=self.self_evm_address,
```

File: src/kakarot/execution_context.cairo [hunk 7]
```diff
@@ -180,6 +186,7 @@ namespace ExecutionContext {
         bitwise_ptr: BitwiseBuiltin*,
     }(self: model.ExecutionContext*, new_memory: model.Memory*) -> model.ExecutionContext* {
         return new model.ExecutionContext(
+            current_contract_address=self.current_contract_address,
```
Comment (LucasLvy) [inline src/kakarot/execution_context.cairo:189]:
```suggestion
            self_sn_address=self.self_sn_address,
            self_evm_address=self.self_evm_address,
```

File: src/kakarot/execution_context.cairo [hunk 8]
```diff
@@ -205,6 +212,7 @@ namespace ExecutionContext {
         self: model.ExecutionContext*, inc_value: felt
     ) -> model.ExecutionContext* {
         return new model.ExecutionContext(
+            current_contract_address=self.current_contract_address,
```
Comment (LucasLvy) [inline src/kakarot/execution_context.cairo:215]:
```suggestion
            self_sn_address=self.self_sn_address,
            self_evm_address=self.self_evm_address,
```

File: src/kakarot/execution_context.cairo [hunk 9]
```diff
@@ -230,6 +238,7 @@ namespace ExecutionContext {
...
```
Comment (LucasLvy) [inline src/kakarot/execution_context.cairo:241]:
```suggestion
            self_sn_address=self.self_sn_address,
            self_evm_address=self.self_evm_address,
```

File: src/kakarot/execution_context.cairo [hunk 10]
```diff
@@ -313,6 +322,7 @@ namespace ExecutionContext {
         check_jumpdest(self, new_pc_offset);
 
         return new model.ExecutionContext(
+            current_contract_address=self.current_contract_address,
```
Comment (LucasLvy) [inline src/kakarot/execution_context.cairo:325]:
```suggestion
            self_sn_address=self.self_sn_address,
            self_evm_address=self.self_evm_address,
```

File: src/kakarot/instructions/block_information.cairo [hunk 11]
```diff
@@ -187,35 +199,75 @@ namespace BlockInformation {
...
+        bitwise_ptr: BitwiseBuiltin*,
+    }(ctx: model.ExecutionContext*) -> model.ExecutionContext* {
+        %{
+            import logging
+            logging.info("0x47 - SELFBALANCE")
+        %}
+        
+        // Get current executing contract address.
+        let current_contract_address: felt = Helpers.uint256_to_felt([ctx.current_contract_address]);
+        let (registry_address_) = registry_address.read();
```
Comment (LucasLvy) [inline src/kakarot/instructions/block_information.cairo:253]:
```suggestion
```
no need to map the address anymore

File: src/kakarot/instructions/block_information.cairo [hunk 12]
```diff
@@ -187,35 +199,75 @@ namespace BlockInformation {
...
+        let (starknet_address) = IRegistry.get_starknet_address(
+            contract_address=registry_address_, evm_address=current_contract_address
+        );
+
+        // Get current executing contract address balance.
+        let (native_token_address_) = native_token_address.read();
+        let (balance: Uint256) = IEth.balanceOf(
+            contract_address=native_token_address_, account=starknet_address
```
Comment (LucasLvy) [inline src/kakarot/instructions/block_information.cairo:261]:
```suggestion
            contract_address=native_token_address_, account=ctx.sn_address
```

File: src/kakarot/kakarot.cairo [hunk 13]
```diff
@@ -22,11 +22,21 @@ func constructor{
 @external
 func execute{
     syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr, bitwise_ptr: BitwiseBuiltin*
-}(code_len: felt, code: felt*, calldata_len: felt, calldata: felt*) -> (
-    stack_len: felt, stack: Uint256*, memory_len: felt, memory: felt*
-) {
+}(
+    current_contract_address_len: felt,
```
Comment (LucasLvy) [inline src/kakarot/kakarot.cairo:26]:
```suggestion
    sn_contract_address: felt,
```

File: src/kakarot/kakarot.cairo [hunk 14]
```diff
@@ -22,11 +22,21 @@ func constructor{
...
+    current_contract_address: Uint256*,
```
Comment (LucasLvy) [inline src/kakarot/kakarot.cairo:27]:
```suggestion
```

File: src/kakarot/kakarot.cairo [hunk 15]
```diff
@@ -22,11 +22,21 @@ func constructor{
...
+    code_len: felt,
+    code: felt*,
+    calldata_len: felt,
+    calldata: felt*,
+) -> (stack_len: felt, stack: Uint256*, memory_len: felt, memory: felt*) {
     alloc_locals;
-    let context = Kakarot.execute(code=code, code_len=code_len, calldata=calldata);
+    let context = Kakarot.execute(
+        current_contract_address=current_contract_address,
```
Comment (LucasLvy) [inline src/kakarot/kakarot.cairo:35]: map sn_address to evm address and then
```suggestion
            self_sn_address=self.self_sn_address,
            self_evm_address=self.self_evm_address,
```

File: src/kakarot/library.cairo [hunk 16]
```diff
@@ -43,7 +44,9 @@ namespace Kakarot {
         pedersen_ptr: HashBuiltin*,
         range_check_ptr,
         bitwise_ptr: BitwiseBuiltin*,
-    }(code: felt*, code_len: felt, calldata: felt*) -> model.ExecutionContext* {
+    }(
+        current_contract_address: Uint256*, code: felt*, code_len: felt, calldata: felt*
```
Comment (LucasLvy) [inline src/kakarot/library.cairo:48]:
```suggestion
        sn_contract_address: felt, code: felt*, code_len: felt, calldata: felt*
```

File: tests/cairo_files/test_execution_context.cairo [hunk 17]
```diff
@@ -33,9 +33,10 @@ func test__init__should_return_an_empty_execution_context{
     tempvar code_len = 1;
     let (calldata) = alloc();
```
Comment (LucasLvy) [inline tests/cairo_files/test_execution_context.cairo:34]: same changes as everywhere

**Review by LucasLvy — CHANGES_REQUESTED** [review]

**Comment by LucasLvy** [comment]
you probably don't have to do the changes everywhere but do it so we only have to map the address once

**Comment by ftupas** [comment]
> you probably don't have to do the changes everywhere but do it so we only have to map the address once

Thanks for the review @LucasLvy 🙂 got it will do that so mapping is done only once

File: src/kakarot/execution_context.cairo [hunk 18]
```diff
@@ -32,7 +31,13 @@ namespace ExecutionContext {
         pedersen_ptr: HashBuiltin*,
         range_check_ptr,
         bitwise_ptr: BitwiseBuiltin*,
-    }(code: felt*, code_len: felt, calldata: felt*) -> model.ExecutionContext* {
+    }(
+        sn_contract_address: felt,
```
Comment (Eikix) [inline src/kakarot/execution_context.cairo:35]: what does sn stand for? starknet?

**Review by Eikix — CHANGES_REQUESTED** [review]
Hey @ftupas :), first of all thanks for your patience, you waited a long time for this PR to be merged
can you explain a bit why we need to add two inputs to the ExecutionContext? I did not understand your intent.

What I understood:
- you're hacking at Selfbalance, which is supposed to return the currently executing contract's own balance in native token
- why do you need to keep track of the starknet address of that contract as well as its evm representation in the execution context?

**Comment by ftupas** [comment]
> Hey @ftupas :), first of all thanks for your patience, you waited a long time for this PR to be merged can you explain a bit why we need to add two inputs to the ExecutionContext? I did not understand your intent.
> 
> What I understood:
> 
>     * you're hacking at Selfbalance, which is supposed to return the currently executing contract's own balance in native token
> 
>     * why do you need to keep track of the starknet address of that contract as well as its evm representation in the execution context?

Hi there @Eikix , thanks for the review and feedback! Prior to the changes from the hackathon, my idea was to include both the starknet and evm contract addresses to the ExecutionContext so the mapping is done once and can be used by the opcodes i.e. getting the native token balance of the executing starknet contract address (mapped from the evm address).  However, I'm redoing this PR right now since the evm and starknet address mapping has been implemented [here](https://github.com/sayajin-labs/kakarot/blob/main/src/kakarot/execution_context.cairo#L100-L102).

**Comment by Eikix** [comment]
> > Hey @ftupas :), first of all thanks for your patience, you waited a long time for this PR to be merged can you explain a bit why we need to add two inputs to the ExecutionContext? I did not understand your intent.
> > What I understood:
> > ```
> > * you're hacking at Selfbalance, which is supposed to return the currently executing contract's own balance in native token
> > 
> > * why do you need to keep track of the starknet address of that contract as well as its evm representation in the execution context?
> > ```
> 
> Hi there @Eikix , thanks for the review and feedback! Prior to the changes from the hackathon, my idea was to include both the starknet and evm contract addresses to the ExecutionContext so the mapping is done once and can be used by the opcodes i.e. getting the native token balance of the executing starknet contract address (mapped from the evm address). However, I'm redoing this PR right now since the evm and starknet address mapping has been implemented [here](https://github.com/sayajin-labs/kakarot/blob/main/src/kakarot/execution_context.cairo#L100-L102).

Hey @ftupas, thanks for your quick reply. I understand, I'd like to apologize for your troubles, we could've coordinated better.

Are you on the kakarot telegram group? If not, kindly provide your telegram handle and I'll add you. This way we'll make sure coordination and communication improves. I want to make sure this is the last time you have to go through all these hardships to merge a PR. Thanks again for your patience 🔥 🚀

**Comment by Eikix** [comment]
> 

Hey @ftupas , after discussing with Lucas, you can still store starknet_contract_address and evm_contract_address in the execution context:)
Tell me when you're done with the changes you wanted to do, but as far as I'm concerned, your PR can be merged

File: src/kakarot/execution_context.cairo [hunk 19]
```diff
@@ -46,6 +51,8 @@ namespace ExecutionContext {
         let memory: model.Memory* = Memory.init();
 
         local ctx: model.ExecutionContext* = new model.ExecutionContext(
+            sn_address=sn_contract_address,
```
Comment (Eikix) [inline src/kakarot/execution_context.cairo:54]: could we rename this starknet_contract_address or starknet_address to keep consistency in the repo?

**Comment by ftupas** [comment]
> > > Hey @ftupas :), first of all thanks for your patience, you waited a long time for this PR to be merged can you explain a bit why we need to add two inputs to the ExecutionContext? I did not understand your intent.
> > > What I understood:
> > > ```
> > > * you're hacking at Selfbalance, which is supposed to return the currently executing contract's own balance in native token
> > > 
> > > * why do you need to keep track of the starknet address of that contract as well as its evm representation in the execution context?
> > > ```
> > 
> > 
> > Hi there @Eikix , thanks for the review and feedback! Prior to the changes from the hackathon, my idea was to include both the starknet and evm contract addresses to the ExecutionContext so the mapping is done once and can be used by the opcodes i.e. getting the native token balance of the executing starknet contract address (mapped from the evm address). However, I'm redoing this PR right now since the evm and starknet address mapping has been implemented [here](https://github.com/sayajin-labs/kakarot/blob/main/src/kakarot/execution_context.cairo#L100-L102).
> 
> Hey @ftupas, thanks for your quick reply. I understand, I'd like to apologize for your troubles, we could've coordinated better.
> 
> Are you on the kakarot telegram group? If not, kindly provide your telegram handle and I'll add you. This way we'll make sure coordination and communication improves. I want to make sure this is the last time you have to go through all these hardships to merge a PR. Thanks again for your patience 🔥 🚀

Thanks for your time to discuss and adding me to the telegram group @Eikix ! As discussed, I redid the PR to take into account the changes to way `ExecutionContext` is initialized (`init` and `init_at_address`). I've marked the test as skip for now and will unmark it once the testing architecture is changed 👍

Appreciate your feedback and reviews!

File: tests/test_zk_evm.py [hunk 20]
```diff
@@ -643,6 +643,17 @@ async def set_account_registry(zk_evm, account_registry):
...
+    {
+        "params": {
+            "code": "4700",
+            "calldata": "",
+            "stack": "420",
+            "memory": "",
+            "return_value": "",
+        },
+        "id": "Get balance of currently executing contract - 0x47 SELFBALANCE",
+        "marks": pytest.mark.execute_at_address,
```
Comment (ftupas) [inline tests/test_zk_evm.py:655]: Test cases with this mark will be tested with `test_execute_at_address`

**Review by Eikix — APPROVED** [review]

**Review by LucasLvy — APPROVED** [review]
didn't check

**Pull request merged**